### PR TITLE
Reintroduce build args for source URLs in Dockerfiles

### PR DESCRIPTION
**Pull request by yhtang**
This is a replacement of #532.

Summary of changes:
- `ci.yaml` now takes an workflow_dispatch input `SOURCE_OVERRIDES` for specifying alternative origin for fetching the source code of the software that we build. The arg takes the form of a comma-separated PACKAGE=URL#REF list to override sources used by build, where PACKAGE∊{JAX,XLA,Flax,transformer-engine,T5X,paxml,praxis,maxtext,levanter,haliax,grok,mujuco,mujuco-mpc} (case-insensitive).
  for example:
  ```
  XLA=https://github.com/yhtang/xla.git#main,transformer-engine=https://github.com/yhtang/TransformerEngine.git#main
  ```
- All dockerfiles now takes in build args of form `URLREF_<SOFTWARE>=<URL>#<GIT-REF>`. These build args have default values that points to the upstream main/master branches.
- Removed `get-source.sh` and replaced it with `git-clone.sh`, which does not implicitly depend on `manifest.yaml`. Instead, it's git clone behavior is completely specified by its arguments. In addition, it will log/append the software checked out to `/opt/manifest.d/git-clone.yaml`. This is in preparation for the next stage of work where we will stop pinning versions and go back to running the presubmit on the head of everything.

Note:
- Triton build is disabled due to the intricacy of specifying default URLREF values per discussion with @olupton; this is re-enabled in https://github.com/NVIDIA/JAX-Toolbox/pull/772

Closes #532

### CODE REVIEW

File: .github/workflows/ci.yaml
Comment (olupton) [inline .github/workflows/ci.yaml:158]: IIUC it would just take renaming a few keys (that contain `-`) in `manifest.yaml` and we could consistently use `_` delimiters everywhere?

**Review by olupton — COMMENTED** [review]
Generally LGTM, a few minor things to be fixed before merging.

File: .github/container/git-clone.sh
Comment (olupton) [inline .github/container/git-clone.sh:98]: This also doesn't match the `manifest.yaml` format, `ref` vs. `tracking_ref` and `commit` vs. `latest_verified_commit`.

**Comment by olupton** [comment]
https://github.com/NVIDIA/JAX-Toolbox/blob/330c61b4c0246ee41a165d22e767a059b7226291/.github/container/Dockerfile.t5x.arm64 also still uses `get-source.sh`.

File: .github/workflows/ci.yaml
Comment (olupton) [reply .github/workflows/ci.yaml:158]: https://github.com/NVIDIA/JAX-Toolbox/blob/330c61b4c0246ee41a165d22e767a059b7226291/.github/container/pip-finalize.sh#L17 is the reason this cannot be done, I think

File: .github/container/git-clone.sh
Comment (olupton) [reply .github/container/git-clone.sh]: I made the output format match the input format, even though "latest verified" doesn't quite make sense. Also changed the key from e.g. `/opt/transformer-engine` to `transformer-engine`, again to try and have only one YAML format.

**Comment by olupton** [comment]
> https://github.com/NVIDIA/JAX-Toolbox/blob/330c61b4c0246ee41a165d22e767a059b7226291/.github/container/Dockerfile.t5x.arm64 also still uses `get-source.sh`.

Changed in 0365a79.

**Review by olupton — DISMISSED** [review]
@terrykong please take a look. You might also like to review the diff between @yhtang's last commit and the current HEAD of the branch.

File: .github/workflows/_ci.yaml
Comment (terrykong) [inline .github/workflows/_ci.yaml:21]: nit: Should the description include that these are case-sensitive?

**Review by terrykong — DISMISSED** [review]
Thanks for reintroducing this feature

It looks like the presubmit shows that all rosetta builds would fail b/c they depended on the [MANIFEST_FILE](https://github.com/NVIDIA/JAX-Toolbox/blob/main/rosetta/Dockerfile.t5x#L29) env var being set. Can that be included in this change-set and make sure at least rosetta-pax succeeds?

Also, upstream-t5x fails to build b/c [optax](https://github.com/NVIDIA/JAX-Toolbox/actions/runs/8879805167/job/24379146824#step:12:213); not due to this PR, but because optax top-of-tree requires `>=0.4.26` and `0.4.26.dev*` does not satisfy. This error would also be present in JAX-Toolbox main so I don't see a reason it'd block this PR

**Comment by yhtang** [comment]
> https://github.com/NVIDIA/JAX-Toolbox/blob/330c61b4c0246ee41a165d22e767a059b7226291/.github/container/Dockerfile.t5x.arm64 also still uses `get-source.sh`.

I left this out intentionally so as to not mess with Georg's ongoing PR.

**Comment by terrykong** [comment]
> Yeah we need to fix that. Can we just hard code the path since we will need to rework that anyway in the follow-up PR?

@ko3n1g put together this PR to fix it: https://github.com/NVIDIA/JAX-Toolbox/pull/773/files

If it looks okay to you, please go ahead and merge

File: .github/container/Dockerfile.base
Comment (DwarKapex) [inline .github/container/Dockerfile.base:203]: Also we need sort of flag whether to apply patch or not to a custom source. Let's say, you wanna build `flax` from custom repo, but you don't want to apply our patches.

File: .github/container/Dockerfile.grok
Comment (DwarKapex) [inline .github/container/Dockerfile.grok:5]: Why `_1`?

File: .github/container/Dockerfile.jax
Comment (DwarKapex) [inline .github/container/Dockerfile.jax:23]: Maybe, add `SRC_PATH_TRANSFORMER_ENGINE`?

File: .github/container/git-clone.sh
Comment (DwarKapex) [inline .github/container/git-clone.sh:5]: Update the `usage()`

File: .github/container/git-clone.sh
Comment (DwarKapex) [inline .github/container/git-clone.sh:32]: Update or remove `while-loop`

File: .github/container/Dockerfile.base
Comment (yhtang) [reply .github/container/Dockerfile.base:203]: The patches are only applied when building the Rosetta container. Maybe @terrykong can answer if this is an immediate need or not.

File: .github/container/Dockerfile.grok
Comment (yhtang) [reply .github/container/Dockerfile.grok:5]: This is the Grok-1 model and there may be future models named Grok-2, Grok-3, etc... 🤣

File: .github/container/git-clone.sh
Comment (yhtang) [reply .github/container/git-clone.sh:5]: What do we need to update here?

File: .github/container/git-clone.sh
Comment (yhtang) [reply .github/container/git-clone.sh:32]: Why shall we remove the while loop?

File: .github/container/git-clone.sh
Comment (yhtang) [reply .github/container/git-clone.sh:79]: Yeah we can make this change when the situation arises.

File: .github/container/pip-finalize.sh
Comment (yhtang) [reply .github/container/pip-finalize.sh:17]: For pip-installed VCS packages, this is still the mechanism in play.

File: .github/workflows/_ci.yaml
Comment (yhtang) [reply .github/workflows/_ci.yaml:21]: Unlike `ci.yaml`, `_ci.yaml` is an internal workflow so I'm not too worried. Let me know if you think otherwise :grin:

**Review by olupton — APPROVED** [review]

File: .github/container/Dockerfile.base
Comment (DwarKapex) [reply .github/container/Dockerfile.base:204]: @terrykong If we apply patches only for Rosetta, maybe, let's move it to rosetta docker files? WDYT?

**Review by DwarKapex — APPROVED** [review]
LGTM after Terry's comments on `patch` folder

File: .github/container/Dockerfile.base
Comment (olupton) [reply .github/container/Dockerfile.base:204]: Moving rosetta-specific stuff out of the base container sounds reasonable to me, but I don't think it's directly linked to this PR so I'm marking this resolved.

**Pull request merged**
